[WIP] - Add methods to CAS Management

SilvestriStefano opened the pull request:
Hello,
I noticed that some methods are missing from the CAS Management service. I am thinking of adding the following for starters

- [ ] load or unload a table
- [ ] save a table 
- [ ] change scope of table 

with related tests as well.

Any suggestion/feedback is highly appreciated.


_Signed-off-by: Stefano Silvestri ssilvestri@olab-studio.com_

smlindauer commented:
Hey @SilvestriStefano,

We do not use f-strings since we still support users back to Python 3.5 and f-strings were introduced in Python 3.6.

I'm opening up the merge request to the git actions now, so those should begin to run momentarily.

If this is a WIP, would it be better to merge your updates into a new branch?

SilvestriStefano commented:
Hello,
I put the WIP tag because I haven't finished writing all the methods. If you prefer we can merge to a different branch.

SilvestriStefano commented:
Hi,
I have added all the functionality that I set out to do. I had to modify the `upload_table()` method because it was not working for me.

Originally, in the post request was written `files={name: file}` which I changed to `files={'files':(name,file)}`. I followed the instructions from the [`request` documentation](https://requests.readthedocs.io/en/latest/user/quickstart/#post-a-multipart-encoded-file). I have also added an extra optional keyword argument which allows for more details to be passed in the body of the request.

I have not added a test method for the `update_state()` method as it is used and tested in other tests. Also I do not have permissions to promote or change state on tables in the `Samples` caslib. This means I would need to upload the table first in some other caslib and then change the state, which is exactly what I do in the `test_upload_file()` method.

smlindauer commented:
Hey @SilvestriStefano,

This looks great! I am in the process of reformatting the sasctl code base to fully follow PEP8 guidelines, so I made a number of comments in my review focused on those reformatting instances. The only logic change I suggested was with regards to adding a new function to handle the many key checks.

Thank you for creating the integration tests and their associated cassettes. I also noticed that cas_management is not subject to any unit tests, so I will add that to my to do list.

SilvestriStefano commented:
Hello @smlindauer,
I ll make all the suggested changes.

Thank you for reviewing.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -8,6 +8,10 @@
+
 DEFAULT_SERVER = "cas-shared-default"
 DEFAULT_CASLIB = "casuser"
 
```
As requested, I have added two functions that handles the many checks: one for checking that the keys entered are valid, and one that checks that required keys are being passed.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        server : str, optional
+            Server where the `caslib` is registered.
+            Defaults to `cas-shared-default`.
+        qparams: dict, optional
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        else:
+            raise ValueError(
+                "The only acceptable query parameters are %s and must be passed in a dictionary"
+                % (allowedQ)
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                    "Missing required query parameters `sourceTableName`, `quiet`, `removeAcs`"
+                )
+            else:
+                query = qParam
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            if ("sourceTableName" not in qParam.keys() 
+            and "quiet" not in qParam.keys() 
+            and "removeAcs" not in qParam.keys()):
+                raise Exception(
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        """
+        server = server or DEFAULT_SERVER
+
+        if qparams is not None:
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+
+        Params
+        ------
+        queryParams : dict, optional
```
done

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
 
     @classmethod
-    def list_caslibs(cls, server, filter_=None):
+    def list_sessions(cls, qparams: dict = None, server: str = None):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        """
+        Returns a collection of sessions available on the CAS server.
+
+        Params
```
my bad. It should be fixed now.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -8,6 +8,10 @@
 from .service import Service
 
+
+from typing import Union, TextIO
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        server = server or DEFAULT_SERVER
+
+        if qparams is not None:
+            allowedQuery = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+                "sessionId",
+            ]
+
+            if not all(key in allowedQuery for key in qparams.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+
+            if not all(key in allowedQuery for key in qparams.keys()):
+                raise ValueError(
+                    "The only acceptable queries are %s." % (allowedQuery)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+                    "The only acceptable queries are %s." % (allowedQuery)
+                )
+            else:
+                query = qparams
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+    def create_session(cls, properties: dict, server: str = None):
+        """Creates a new session on the CAS server.
+
+        Params
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        """
+        server = server or DEFAULT_SERVER
+
+        allowedBodyKeys = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+            "timeOut",
+        ]
+
+        if not all(key in allowedBodyKeys for key in properties.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+
+        if not all(key in allowedBodyKeys for key in properties.keys()):
+            raise ValueError(
+                "The only acceptable properties are %s." % (allowedBodyKeys)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        return sess
+
+    @classmethod
+    def delete_session(cls, sess_id: str, server: str = None, qparams: dict = None):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+            A string indicating the Session id.
+        server : str
+            Name of the CAS server.  Defaults to `cas-shared-default`.
+        qparams : dict, optional
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+    def delete_session(cls, sess_id: str, server: str = None, qparams: dict = None):
+        """Terminates a session on the CAS server.
+
+        Params
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        """
+        server = server or DEFAULT_SERVER
+
+        if qparams is not None:
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        server = server or DEFAULT_SERVER
+
+        if qparams is not None:
+            allowedQueryKeys = ["force", "superUserSessionId"]
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        if qparams is not None:
+            allowedQueryKeys = ["force", "superUserSessionId"]
+
+            if not all(key in allowedQueryKeys for key in qparams.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+
+            if not all(key in allowedQueryKeys for key in qparams.keys()):
+                raise ValueError(
+                    "The only acceptable queries are %s." % (allowedQueryKeys)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+                    "The only acceptable queries are %s." % (allowedQueryKeys)
+                )
+        else:
+            qparams = {}
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        else:
+            qparams = {}
+
+        sess = cls.delete("/servers/%s/sessions/%s" % (server, sess_id), params=qparams)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
         if format_ is not None:
             data["format"] = format_
 
+        allowedBody = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            'stringLengthMultiplier', 
+            'varcharConversionThreshold'
+        ]
+        allowedBcsv = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            'threadCount', 
+            'stripBlanks'
+        ]
+        allowedBxls = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            'replace', 
+            'sheetName'
+        ]
+        allowedBsas = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ]
+
+        if detail is not None:
+            inputK = detail.keys()
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+
+        if detail is not None:
+            inputK = detail.keys()
+            if not all(key in allowedBody for key in inputK):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            inputK = detail.keys()
+            if not all(key in allowedBody for key in inputK):
+                raise ValueError(
+                    "The body accepts only the following parameters %s." % (allowedBody)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                )
+            elif (
+                format_ == 'csv' and 
+                not all(key in allowedBcsv for key in inputK)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                not all(key in allowedBcsv for key in inputK)
+                ):
+                    raise ValueError(
+                        "The parameters specific to a csv file are %s." % (allowedBcsv)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                    )
+            elif (
+                format_ in ['xls','xlsx'] and
+                not all(key in allowedBxls for key in inputK)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                not all(key in allowedBxls for key in inputK)
+                ):
+                    raise ValueError(
+                        "The parameters specific to a excel file are %s." % (allowedBxls)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                    )
+            elif (
+                format_ in ['sashdat','sas7bdat'] and
+                not all(key in allowedBsas for key in inputK)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                not all(key in allowedBsas for key in inputK)
+                ):
+                    raise ValueError(
+                        "The parameters specific to a sas file are %s." % (allowedBsas)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        caslib: str = None,
+        server: str = None,
+        *,
+        qparams: dict = None,
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                "The state can only have values of `loaded` or `unloaded`."
+            )
+
+        if qparams is not None:
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            )
+
+        if qparams is not None:
+            allowedQueryKeys = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                "createRelationships",
+            ]
+
+            if not all(key in allowedQueryKeys for key in qparams.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+
+            if not all(key in allowedQueryKeys for key in qparams.keys()):
+                raise ValueError(
+                    "The only query parameters allowed are %s." % (allowedQueryKeys)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                    "The only query parameters allowed are %s." % (allowedQueryKeys)
+                )
+            else:
+                query.update(qparams)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            else:
+                query.update(qparams)
+
+        allowedBodyKeys = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            "scope",
+        ]
+
+        if not all(key in allowedBodyKeys for key in body.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+
+        if not all(key in allowedBodyKeys for key in body.keys()):
+            raise ValueError(
+                "The body accepts only the following parameters %s." % (allowedBodyKeys)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        server = server or DEFAULT_SERVER
+
+        if properties:
+            allowedBodyKeys = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+                "parameters",
+            ]
+
+            if not all(key in allowedBodyKeys for key in properties.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+
+            if not all(key in allowedBodyKeys for key in properties.keys()):
+                raise ValueError(
+                    "The only acceptable properties are %s." % (allowedBodyKeys)
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+    def del_table(
+        cls, 
+        name: str, 
+        qParam: dict = None, 
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ----------
+        name : str
+            Name of the table.
+        qParam : dict
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        server = server or DEFAULT_SERVER
+        caslib = caslib or DEFAULT_CASLIB
+        
+        allowedQ = [
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            "removeAcs"
+        ]
+
+        if isinstance(qParam,dict) and all(key in allowedQ for key in qParam.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ]
+
+        if isinstance(qParam,dict) and all(key in allowedQ for key in qParam.keys()):
+            if ("sourceTableName" not in qParam.keys() 
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+
+        if isinstance(qParam,dict) and all(key in allowedQ for key in qParam.keys()):
+            if ("sourceTableName" not in qParam.keys() 
+            and "quiet" not in qParam.keys() 
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        if isinstance(qParam,dict) and all(key in allowedQ for key in qParam.keys()):
+            if ("sourceTableName" not in qParam.keys() 
+            and "quiet" not in qParam.keys() 
+            and "removeAcs" not in qParam.keys()):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        name : str
+            Name of the table.
+        qParam : dict
+            Query parameters. Note that some are required.
```
fixed.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        else:
+            properties = {}
+
+        query = {"sessionId": sessId} if sessId else {}
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        name: str,
+        caslib: str,
+        properties: dict = None,
+        sessId: str = None,
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ----------
+        name : str
+            Name of the table.
+        sessId: str
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+            The session ID
+        caslib : str
+            Name of the caslib.
+        propertes : dict, optional
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        """
+        server = server or DEFAULT_SERVER
+
+        query = {"value": "global", "sessionId": sessId}
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ----------
+        name : str
+            Name of the table.
+        sessId: str
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        return tbl
+
+    @classmethod
+    def promote_table(cls, name: str, sessId: str, caslib: str, server: str = None):
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -178,9 +317,336 @@ def upload_file(
+        ----------
+        value : str, required
+            State to which to set the table. Valid values include `loaded` or `unloaded`.
+        name : str, required
```
fixed.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+            query={}
+        
+        sesslist = cls.get(
+            "/servers/%s/sessions"  % (server),
```
done.

SilvestriStefano commented on src/sasctl/_services/cas_management.py:
```diff
@@ -22,7 +26,125 @@ class CASManagement(Service):
+        else:
+            query={}
+        
+        sesslist = cls.get(
```
done.

SilvestriStefano commented:
I want to mention that I changed the `DEFAULT_CASLIB` value to `Samples` since `casuser` is not a valid caslib. The caslib associated to a user is `CASUSER(user_id)` were `user_id` is the id of the user (for example in my integration tests I have used `CASUSER(X1064007)` )

smlindauer commented:
> I want to mention that I changed the `DEFAULT_CASLIB` value to `Samples` since `casuser` is not a valid caslib. The caslib associated to a user is `CASUSER(user_id)` were `user_id` is the id of the user (for example in my integration tests I have used `CASUSER(X1064007)` )

Should we go with Samples or would Public make more sense? I believe Public and Samples are always created, but Public is more likely to contain CAS data that users would be using.

SilvestriStefano commented:
@smlindauer  you are right. The "Public" caslib makes more sense as the default one. I made the change.
